### PR TITLE
Register new routes via SDK as named routes

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
Continuation of #622 
This change allows registering extension routes as NamedRoute (new, deprecated, replaced), and updates sample helloworld extension to conform to the new route registration mechanism to optionally add legacy permission actions (i.e. `cluster:admin/opensearch/.....`)

Related PRs:
- Blocked by: https://github.com/opensearch-project/OpenSearch/pull/7957
- https://github.com/opensearch-project/security/pull/2753

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by DarshitChanpura** [comment]
Gradle check is failing as it needs core artifact with changes from this: https://github.com/opensearch-project/OpenSearch/pull/7957 merged

File: DEVELOPER_GUIDE.md
Comment (dbwiddis) [inline DEVELOPER_GUIDE.md:177]:
```suggestion
Note: If the Security plugin is initialized in OpenSearch, use admin credentials to send the extension initialization request.
```

**Review by dbwiddis — COMMENTED** [review]
LGTM with a few documentation nits. 

Also didn't comment inline, but I'm not sure rewriting temporary Deprecated and Replaced route handlers is the right approach.  I understand we will eventually incorporate this functionality into the NamedRoutes, correct?

**Comment by DarshitChanpura** [comment]
@dbwiddis I've removed references to named handlers for deprecated and replaced routes.

**Review by dbwiddis — DISMISSED** [review]

File: PLUGIN_MIGRATION.md
Comment (cwperks) [inline PLUGIN_MIGRATION.md:97]: Do you think its worth mentioning here that `actionNames` are useful for plugins being converted to extensions and wanting to keep backward compatibility with existing reserved roles that they have defined?

File: src/main/resources/sample/helloworld-settings.yml
Comment (cwperks) [inline src/main/resources/sample/helloworld-settings.yml:2]: Is `shortExtensionName` being used? I know we had talked about a separate values for extension unique id and an abbreviation for an extension, but is the extension unique id now also being treated as an abbreviation that would be prefixed to all action names? i.e. The route `_extensions/_hw/greet` is assigned the permissionable action name `hw:greet` and `hw` is the extension unique id

File: src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java
Comment (cwperks) [inline src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java:202]: Is it enforced anywhere that `actionNames` cannot contain spaces? If space is being used as a delimiter than an `actionName` must not contain a space

File: src/main/resources/sample/helloworld-settings.yml
Comment (cwperks) [reply src/main/resources/sample/helloworld-settings.yml:2]: Ok, I see now that extension name !== extension unique id. That sounds fine to me.

**Review by cwperks — COMMENTED** [review]
Thank you @DarshitChanpura . Took a pass and this looks good to me. There's one reference to `shortExtensionName` that looks like it can be removed.

File: src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java
Comment (cwperks) [inline src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java:69]: One last comment, it looks like an extensions developer has the ability to pass any name as unique name and conventionally they would opt to use this `routePrefix`. I'm wondering if it should be stronger than that and that all names for routes with an extension must be prefixed with the extension's name/abbreviation?

**Review by cwperks — APPROVED** [review]
This PR looks good to me. I left a few comments in a review earlier, but none that require to be addressed before this PR is merged.

File: src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java:202]: I had this same question and [yes, it is](https://github.com/DarshitChanpura/OpenSearch/blob/bea26f97c66070581070764fe4790d2f0579c8e7/server/src/main/java/org/opensearch/rest/NamedRoute.java#L27).

File: CREATE_YOUR_FIRST_EXTENSION.md
Comment (dbwiddis) [inline CREATE_YOUR_FIRST_EXTENSION.md:167]: This needs to be `public` because of the public method it's inheriting.  Don't hold up PR approval for this, I'll address this in a follow-up doc change.

File: src/main/resources/sample/helloworld-settings.yml
Comment (DarshitChanpura) [reply src/main/resources/sample/helloworld-settings.yml]: removed the reference.

File: CREATE_YOUR_FIRST_EXTENSION.md
Comment (DarshitChanpura) [reply CREATE_YOUR_FIRST_EXTENSION.md]: Done.

File: PLUGIN_MIGRATION.md
Comment (DarshitChanpura) [reply PLUGIN_MIGRATION.md:97]: done.

File: src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java:69]: I think we should leave this responsibility to extensions developer. We can maybe in a future iteration enforce strict checks to validate the prefixes with extension name.

**Comment by DarshitChanpura** [comment]
@dbwiddis Would you please re-review? 

The build seems to broken due to this change: https://github.com/opensearch-project/OpenSearch/commit/52a5e3f6e0ca599e3193807134ea42660ecdd195

**Comment by DarshitChanpura** [comment]
> @dbwiddis Would you please re-review?
> 
> The build seems to broken due to this change: [opensearch-project/OpenSearch@52a5e3f](https://github.com/opensearch-project/OpenSearch/commit/52a5e3f6e0ca599e3193807134ea42660ecdd195)

Added a fix.

**Comment by dbwiddis** [comment]
> > @dbwiddis Would you please re-review?
> > The build seems to broken due to this change: [opensearch-project/OpenSearch@52a5e3f](https://github.com/opensearch-project/OpenSearch/commit/52a5e3f6e0ca599e3193807134ea42660ecdd195)
> 
> Added a fix.

Yep, I'll re-review and thank you for fixing that!  I was playing around off and on fixing up the sample extension that this documents, and encountered that same problem. I just deleted the offending lines. ;)

BTW, updated sample at https://github.com/dbwiddis/CRUDExtension I am working on updating the docs starting with your version in this PR

**Review by dbwiddis — DISMISSED** [review]
LGTM.  

I have updated the "first extension" that the CREATE_YOUR_FIRST_EXTENSION doc describes and will be updating that doc file in a separate PR shortly.   See https://github.com/dbwiddis/CRUDExtension

File: build.gradle
Comment (DarshitChanpura) [reply build.gradle]: done.

**Comment by DarshitChanpura** [comment]
@dbwiddis @owaiskazi19 Would you please review this again?

File: src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/rest/ExtensionRestPathRegistry.java:107]: Nit: Instead of passing null here we can have a separate method for  `registerHandler` without `name` and `actionNames.`

**Review by owaiskazi19 — APPROVED** [review]

**Review by dbwiddis — APPROVED** [review]

**Comment by dbwiddis** [comment]
@DarshitChanpura can you please create a backport PR?

**Comment by joshpalis** [comment]
@DarshitChanpura Can you confirm that the hello world extension can be started with these new changes. I am seeing errors when running `./gradlew helloWorld`

```
19:15:52.052 [main] DEBUG org.opensearch.threadpool.ThreadPool - created thread pool: name [write], size [8], queue size [10k]
19:15:52.052 [main] DEBUG org.opensearch.threadpool.ThreadPool - created thread pool: name [snapshot], core [1], max [4], keep alive [5m]
19:15:52.053 [main] DEBUG org.opensearch.threadpool.ThreadPool - created thread pool: name [search_throttled], size [1], queue size [100]
Exception in thread "main" OpenSearchException[Invalid route name specified. The route name may include the following characters 'a-z', 'A-Z', '0-9', ':', '/', '*', '_' and be less than 250 characters]
        at org.opensearch.rest.NamedRoute.<init>(NamedRoute.java:143)
        at org.opensearch.rest.NamedRoute$Builder.build(NamedRoute.java:112)
        at org.opensearch.sdk.sample.helloworld.rest.RestHelloAction.routes(RestHelloAction.java:71)
        at org.opensearch.sdk.rest.ExtensionRestPathRegistry.registerHandler(ExtensionRestPathRegistry.java:43)
        at org.opensearch.sdk.ExtensionsRunner.<init>(ExtensionsRunner.java:240)
        at org.opensearch.sdk.ExtensionsRunner.run(ExtensionsRunner.java:583)
        at org.opensearch.sdk.sample.helloworld.HelloWorldExtension.main(HelloWorldExtension.java:82)
```

**Comment by dbwiddis** [comment]
> @DarshitChanpura Can you confirm that the hello world extension can be started with these new changes. I am seeing errors when running `./gradlew helloWorld`

The problem is the route prefix code:
```java
protected static String routePrefix(String name) {
    return extensionName + ":" + name;
}
```
It uses the extension name without checking it; the hello world extension name (per the .yml file) is 
```yaml
extensionName: hello-world
```
The `-` is not allowed.

**Comment by DarshitChanpura** [comment]
Aah. I've tested it using the short hand `hw` and didn't run into this error. I can create a PR to rename this to `hello_world` which will work.

**Comment by owaiskazi19** [comment]
> Aah. I've tested it using the short hand `hw` and didn't run into this error. I can create a PR to rename this to `hello_world` which will work.

The correct way of adding a route would be `helloWorld` rather with a `-` or `_`. Can you rename it with the same @DarshitChanpura?

**Comment by DarshitChanpura** [comment]
@dbwiddis @joshpalis This PR should resolve the issue. https://github.com/opensearch-project/opensearch-sdk-java/pull/868

**Pull request merged**
